### PR TITLE
Update pooch download, rename model_type->model_name in get_sam_model

### PR DESCRIPTION
**Pull request by constantinpape**
This PR contains two changes:
- Clean up the model download with pooch.
- Rename the `model_type` argument of `get_sam_model` to `model_name`. (This had to be changed in a few files, that's why the changelog is rather big.)

@GenevieveBuckley I left a few comments on what the effect of passing a checkpoint via `checkpoint_path` is.
In a nutshell:
- `checkpoint_path is None`: we download the "known" model corresponding to `model_name` via `pooch`. 
- `checkpoint_path is not None`: we use the model file stored locally (a URL is not allowed here!). These weights are NOT copied to the cache directory or compared to any hashes. They are just used as is. (If it's not a valid weight file there will be an error when trying to actually load those weights later).

I hope that is clear from the comments now. (As a sidenote: I understand your idea to split this into several functions, but we would need to change quite a lot of downstream functionality for this, and I also think the current design makes it quite convenient to use models finetuned by the user, which is the main use-case for `checkpoint_path`).

I have checked that this works with the example scripts and we also have a test now that checks it with `checkpoint_path` as an argument.

### CODE REVIEW

File: micro_sam/sam_annotator/_widgets.py [hunk 1]
```diff
@@ -32,7 +31,7 @@ def embedding_widget(
     pbar: widgets.ProgressBar,
     image: "napari.layers.Image",
     model: Literal[tuple(models().urls.keys())] = _DEFAULT_MODEL,
-    device: Literal[tuple(["auto"] + _available_devices())]= "auto",
+    device: Literal[tuple(["auto"] + _available_devices())] = "auto",
```
Comment (constantinpape) [inline micro_sam/sam_annotator/_widgets.py:34]: There were a bunch of PEP8 changes in this file, that's why there is a diff here, the only actual change in here is renaming `model_type->model_name`.

File: micro_sam/sam_annotator/_widgets.py [hunk 2]
```diff
@@ -4,7 +4,6 @@
 
 from magicgui import magic_factory, widgets
 from napari.qt.threading import thread_worker
-import pooch
```
Comment (constantinpape) [inline micro_sam/sam_annotator/_widgets.py:7]: Unused import.

File: micro_sam/util.py [hunk 3]
```diff
@@ -109,35 +113,9 @@ def models():
             "vit_b_em": "https://zenodo.org/record/8250260/files/vit_b_em.pth?download=1",
         },
     )
-    # This extra dictionary is needed for _get_checkpoint to download weights from a specific checkpoint
```
Comment (constantinpape) [inline micro_sam/util.py:112]: We don't need this anymore. We just use the keys in the registry as filenames in the cache.

File: micro_sam/util.py [hunk 4]
```diff
@@ -109,35 +113,9 @@ def models():
...
-            "vit_t": "vit_t_mobile_sam.pth",
-            "vit_h_lm": "vit_h_lm.pth",
-            "vit_b_lm": "vit_b_lm.pth",
-            "vit_h_em": "vit_h_em.pth",
-            "vit_b_em": "vit_b_em.pth",
-        }
     return models
 
 
-def _get_checkpoint(model_name, checkpoint_path):
```
Comment (constantinpape) [inline micro_sam/util.py:126]: This is not needed anymore. We easier use the model downloaded by pooch and stored in the cache directory, or we use a local weight file.

In the second case we do not copy that weight file or compare its hash.

File: micro_sam/util.py [hunk 5]
```diff
@@ -200,16 +178,23 @@ def _available_devices():
...
+    This function will download the required model or load it from the cached weight file.
+    This location of the cache can be changed by setting the environment variable: MICROSAM_CACHEDIR.
+    The name of the requested model can be set via `model_name`.
+    See https://computational-cell-analytics.github.io/micro-sam/micro_sam.html#finetuned-models
+    for an overview of the available models
+
+    Alternatively this function can also load a model from weights stored in a local filepath.
+    The corresponding file path is given via `checkpoint_path`. In this case `model_name`
+    must be given as the matching encoder architecture, e.g. "vit_b" if the weights are for
+    a SAM model with vit_b encoder.
```
Comment (constantinpape) [inline micro_sam/util.py:197]: I added some explanation to try to make clear what happens depending on passing a `checkpoint_path` or not. Let me know if this is still unclear and what could be improved.

**Comment by constantinpape** [comment]
p.s. : it looks like you have a PEP8 checker here that is not active in the main repo. We could also activate it there, but should increase the line length limit to 120, which is more in line with modern python.

**Comment by GenevieveBuckley** [comment]
The PEP8 bot is not deliberate - I think I just enabled it for one repo but somehow it just runs on all of them now. You can safely ignore it.

File: micro_sam/training/util.py [hunk 6]
```diff
@@ -14,12 +14,13 @@ def get_trainable_sam_model(
     device: Optional[Union[str, torch.device]] = None,
     checkpoint_path: Optional[Union[str, os.PathLike]] = None,
     freeze: Optional[List[str]] = None,
-    checkpoint_path: Optional[Union[str, os.PathLike]] = None,
```
Comment (GenevieveBuckley) [inline micro_sam/training/util.py:17]: Are you removing this line on purpose, @constantinpape ?

File: micro_sam/util.py (hunk 5 above)
Comment (GenevieveBuckley) [reply micro_sam/util.py:197]: This helps - I think the key point is that checkpoint_path is only for **local** weights. I was imagining a situation where people would share their finetuned weights and users could download them from zenodo or similar (but I guess that's what the `get_custom_sam_model` function is for)

File: micro_sam/util.py (hunk 4 above)
Comment (GenevieveBuckley) [reply micro_sam/util.py:126]: 👍

**Comment by GenevieveBuckley** [comment]
I'm going to merge this (I'm off work for the next four days, so good to keep things moving even if we need to make more changes later).

1. Can you confirm that you meant to remove the `checkpoint_path` kwarg from the `trainable_sam_model` function in `micro_sam/training/util.py`? If that line is not meant to be deleted, we'll need an edit to restore it.

2. I've changed my mind about one thing. I think it's best if we do not change model_type -> model_name in only parts of the code. It's much simpler to distinguish `model_type_` with a more obviously different variable name (I've chosen `abbreviated_model_type`). This is (1) much less work since it only appears in one function, whereas the model_type kwarg is scattered around everywhere, and (2) doesn't also require us to change other functions with similar kwarg inputs, like `get_custom_sam_model`. 

I know it was originally my suggestion, and you did put extra work in because of this, but I think this is less confusing overall.

3. I really like the rest of this PR - I think it is much neater now, and the extra documentation and explanation is helpful.

**Comment by GenevieveBuckley** [comment]
... ok, I guess I'm not about to merge this yet.
I don't know why the windows test has a failure but the others do not. I have to go now, sorry
```
FAILED test/test_util.py::TestUtil::test_get_sam_model - ValueError: Checkpoint at C:\Users\runneradmin\AppData\Local\micro_sam\micro_sam\Cache\models\vit_t could not be found.
```

FYI I have tried to add you as a collaborator to my fork repo, so hopefully you can merge the PR, or push directly to my open PR on the main micro-sam repo if need be.

File: micro_sam/training/util.py (hunk 6 above)
Comment (constantinpape) [reply micro_sam/training/util.py:17]: Yes. That's on purpose, It somehow has been duplicated (see line 15 there is also the argument `checkpoint_path`). So this is just a small bugfix. (I am not sure how this got introduced, maybe in a recent merge.)

File: micro_sam/util.py (hunk 5 above)
Comment (constantinpape) [reply micro_sam/util.py:197]: > This helps - I think the key point is that checkpoint_path is only for local weights.

:+1:

> I was imagining a situation where people would share their finetuned weights and users could download them from zenodo or similar (but I guess that's what the get_custom_sam_model function is for)

For now the checkpoint would need to be downloaded for this. In the future we want to support uploading models to bioimage.io and this is how we would support uploading and downloading other finetuned models.
(But we are still waiting for some changes on the bioimage.io for this.)

**Comment by constantinpape** [comment]
> 1. Can you confirm that you meant to remove the `checkpoint_path` kwarg from the `trainable_sam_model` function in `micro_sam/training/util.py`? If that line is not meant to be deleted, we'll need an edit to restore it.

Yep, that's correct and was a bugfix. (See comment).

> 2\. I've changed my mind about one thing. I think it's best if we do not change model_type -> model_name in only parts of the code. It's much simpler to distinguish `model_type_` with a more obviously different variable name (I've chosen `abbreviated_model_type`). This is (1) much less work since it only appears in one function, whereas the model_type kwarg is scattered around everywhere, and (2) doesn't also require us to change other functions with similar kwarg inputs, like `get_custom_sam_model`.

Makes sense and no worries! (I did a search and replace on selected files so that was not much work).

> 3\. I really like the rest of this PR - I think it is much neater now, and the extra documentation and explanation is helpful.

Thanks! Overall I really like the changes due to pooch, this removes a lot of code and the whole handling of checkpoints got much easier.

**Comment by constantinpape** [comment]
> ... ok, I guess I'm not about to merge this yet.
> I don't know why the windows test has a failure but the others do not. I have to go now, sorry

Adding me has worked and I can merge. I will look into the test failure and merge it once fixed.

Enjoy your days of!f And I will finally follow up with a longer mail on next steps in the meantime :).

**Comment by constantinpape** [comment]
So the problem was that the `os.pooch_cache` was applied twice. This probably caused some issues with windows file paths.
Everything is passing now and I will merge.

**Pull request merged**